process/smartContract/scQueryService.go

```diff
@@ -116,60 +176,204 @@ func (service *SCQueryService) shouldAllowQueriesExecution() bool {
 		}
 	}
 
 	if query.SameScState {
 		err = service.checkForRootHashChanges(rootHashBeforeExecution)
 		if err != nil {
-			return nil, err
+			return nil, nil, err
+		}
+	}
+
+	var blockHash []byte
+	var blockNonce uint64
+	if !check.IfNil(blockHeader) {
+		blockNonce = blockHeader.GetNonce()
+		blockHash, err = core.CalculateHash(service.marshaller, service.hasher, blockHeader)
+		if err != nil {
+			return nil, nil, err
 		}
 	}
+	blockInfo := holders.NewBlockInfo(blockHash, blockNonce, blockRootHash)
+	return vmOutput, blockInfo, nil
+}
+
+// TODO: extract duplicated code with nodeBlocks.go
+func (service *SCQueryService) extractBlockHeaderAndRootHash(query *process.SCQuery) (data.HeaderHandler, []byte, error) {
+
+	if len(query.BlockHash) > 0 {
+		blockHeader, err := service.getBlockHeaderByHash(query.BlockHash)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		header := service.getBlockHeader(blockHeader)
+		return header, header.GetRootHash(), nil
+	}
+
+	if query.BlockNonce.HasValue {
+		blockHeader, _, err := service.getBlockHeaderByNonce(query.BlockNonce.Value)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		header := service.getBlockHeader(blockHeader)
+		return header, header.GetRootHash(), nil
+	}
+
+	return service.mainBlockChain.GetCurrentBlockHeader(), service.mainBlockChain.GetCurrentBlockRootHash(), nil
+}
+
+func (service *SCQueryService) getBlockHeaderByHash(headerHash []byte) (data.HeaderHandler, error) {
+	epoch, err := service.getOptionalEpochByHash(headerHash)
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := service.getBlockHeaderInEpochByHash(headerHash, epoch)
+	if err != nil {
+		return nil, err
+	}
+
+	return header, nil
+}
+
+func (service *SCQueryService) getOptionalEpochByHash(hash []byte) (core.OptionalUint32, error) {
+	if !service.historyRepository.IsEnabled() {
+		return core.OptionalUint32{}, nil
+	}
+
+	epoch, err := service.historyRepository.GetEpochByHash(hash)
+	if err != nil {
+		return core.OptionalUint32{}, err
+	}
+
+	return core.OptionalUint32{Value: epoch, HasValue: true}, nil
+}
+
+func (service *SCQueryService) getBlockHeaderInEpochByHash(headerHash []byte, epoch core.OptionalUint32) (data.HeaderHandler, error) {
+	shardId := service.shardCoordinator.SelfId()
+	unitType := dataRetriever.GetHeadersDataUnit(shardId)
+	storer, err := service.storageService.GetStorer(unitType)
+	if err != nil {
+		return nil, err
+	}
+
+	var headerBuffer []byte
+
+	if epoch.HasValue {
+		headerBuffer, err = storer.GetFromEpoch(headerHash, epoch.Value)
+	} else {
+		headerBuffer, err = storer.Get(headerHash)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	header, err := process.UnmarshalHeader(shardId, service.marshaller, headerBuffer)
+	if err != nil {
+		return nil, err
+	}
+
+	return header, nil
+}
+
+func (service *SCQueryService) getBlockHeader(currentHeader data.HeaderHandler) data.HeaderHandler {
+	header, _, err := service.getBlockHeaderByNonce(currentHeader.GetNonce() + 1)
```

Comment:
why nonce +1 here?
Maybe rename the caller method to getNextBlockHeader?

also I don't think we need this method, as we already have getBlockByNonce (current header nonce +1), and an error on this get should not use the current header, but propagate the error.

Reply:
removed the method and updated as suggested